factor/grace/classes/factor.php

```diff
@@ -111,9 +111,28 @@ public function get_state() {
             $duration = get_config('factor_grace', 'graceperiod');
 
             if (!empty($duration)) {
-                return (time() <= $starttime + $duration)
-                ? \tool_mfa\plugininfo\factor::STATE_PASS
-                : \tool_mfa\plugininfo\factor::STATE_NEUTRAL;
+                if (time() > $starttime + $duration) {
+                    // If gracemode would have given points, but now doesnt,
+                    // Jump out of the loop and force a factor setup.
+                    // We will return once there is a setup, or the user tries to leave.
+                    if (get_config('factor_grace', 'forcesetup') && !$noredir) {
+                        $redir = true;
+                        $factorurls = \tool_mfa\manager::get_no_redirect_urls();
+                        foreach ($factorurls as $factorurl) {
+                            if ($factorurl->compare($PAGE->url)) {
+                                $redir = false;
+                            }
+                        }
+
+                        if ($redir) {
+                            redirect(new \moodle_url('/admin/tool/mfa/user_preferences.php'),
+                                get_string('redirectsetup', 'factor_grace'));
+                        }
+                    }
+                    return \tool_mfa\plugininfo\factor::STATE_NEUTRAL;
+                } else {
+                    \tool_mfa\plugininfo\factor::STATE_PASS;
```

Comment:
Missing `return` keyword

Reply:
Also how is this not something PHP linting/CS/CI doesn't pick up. A no-op line should be flagged imo.